pkg/client/kubeconfig.go

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package client
+
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"go.uber.org/zap"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/client-go/transport"
+	certutil "k8s.io/client-go/util/cert"
+)
+
+func isKubeConfigStillValid(kubeConfigPath string, logger *zap.Logger) (bool, error) {
+	logger.Debug("checking if kubeconfig exists...")
+
+	_, err := os.Stat(kubeConfigPath)
+	if os.IsNotExist(err) {
+		logger.Debug("kubeconfig does not exist. bootstrapping will continue")
+		return false, nil
+	}
+	if err != nil {
+		logger.Debug("error reading existing bootstrap kubeconfig. bootstrapping will continue", zap.Error(err))
+		return false, nil // not returning an error so bootstrap can continue
+	}
+
+	isValid, err := isClientConfigStillValid(kubeConfigPath, logger)
+	if err != nil {
+		return false, fmt.Errorf("unable to load kubeconfig: %v", err)
+	}
+	if isValid {
+		logger.Debug("kubeconfig is valid. bootstrapping will not continue")
+		return true, nil
+	}
+
+	logger.Debug("kubeconfig is invalid. bootstrapping will continue")
+	return false, nil
+}
+
+// copied from https://github.com/kubernetes/kubernetes/blob/e45f5b089f770b1c8a1583f2792176bfe450bb47/pkg/kubelet/certificate/bootstrap/bootstrap.go#L231
+// isClientConfigStillValid checks the provided kubeconfig to see if it has a valid
+// client certificate. It returns true if the kubeconfig is valid, or an error if bootstrapping
+// should stop immediately.
+func isClientConfigStillValid(kubeconfigPath string, logger *zap.Logger) (bool, error) {
+	bootstrapClientConfig, err := loadRESTClientConfig(kubeconfigPath)
+	if err != nil {
+		logger.Error("unable to read existing bootstrap client config from ",
+			zap.String("kubeconfigPath", kubeconfigPath),
```

Comment:
nit: I think zap.Error/zap.String... etc just adds the given error/string as a dimension, you probably want the message itself to be self-contained - e.g. ("unable to read specified kubeconfig")

I could be wrong on this though

Reply:
same comment for below logger.Error calls

Reply:
the below messages probably also shouldn't refer to "bootstrap client config" - in this case we're talking about the specified kubeconfig